### PR TITLE
Use mock config in tests

### PR DESCRIPTION
**Pull request by mfpierre**
### What does this PR do?

Introduce a config mock object authorizing `set`

### Motivation

Make `Set` private in the config package

### CODE REVIEW

File: pkg/config/mock.go
Comment (gbbr) [inline pkg/config/mock.go]: This confused me initially because it gives the impression that it simply returns a mock config and it's not immediately obvious that it replaces the global one, so I'd suggest:
```go
// Mock replaces the global config with a mock one and returns an interface to allow
// editing it.
func Mock() (*MockConfig)
```
I'm also wondering if we should have a second return parameter which would restore any configuration that was replaced by this call.

File: pkg/config/mock.go
Comment (gbbr) [inline pkg/config/mock.go]: Not sure if this really provides value since we already have `MergeConfig` exposed. I am thinking that most tests (at least for trace) would look like:
```go
config.Mock().MergeConfig(strings.NewReader(`
    api_key: test
    # ... more yaml
`)
```

**Review by gbbr — DISMISSED** [review]
This LGTM. I just have some questions but I don't feel strongly about any of them. This will definitely come in handy.

File: pkg/util/clusteragent/cluster_agent.auth_token
Comment (olivielpeau) [inline pkg/util/clusteragent/cluster_agent.auth_token]: I don't think you meant to add this file to the PR. Where does it come from btw?

File: pkg/config/mock.go
Comment (olivielpeau) [reply pkg/config/mock.go]: +1 on changing the name of the function to `Mock` (unless there's a good reason not to), and mentioning that it replaces the global config

**Review by olivielpeau — DISMISSED** [review]
LGTM overall!

Left a couple of minor comments

File: pkg/util/clusteragent/cluster_agent.auth_token
Comment (mfpierre) [reply pkg/util/clusteragent/cluster_agent.auth_token]: probably a file not cleaned up after unit test 🤔

File: pkg/config/mock.go
Comment (mfpierre) [reply pkg/config/mock.go]: I agree it's clearer 👍

**Review by iksaif — DISMISSED** [review]

**Review by gbbr — DISMISSED** [review]
Would love to hear an answer to [this message](https://github.com/DataDog/datadog-agent/pull/2560#issuecomment-441240199) before considering a merge.

**Comment by mfpierre** [comment]
> If we are keeping Set and the package level exported Datadog identifier, then there is technically no use for this. It can just be overridden on a per-test basis.

Hey @gbbr this PR is only one part of the process, for example there are also #2556 and #2553 that goes into this direction. We choose an incremental approach to break-down the work in severals PRs but once everything is merged we'll definitely make `Set` private.

**Comment by gbbr** [comment]
I've dismissed my PR to unblock this but consider that if we want to make the config read-only there are other methods which may be useful to agent that should potentially be left. Currently the trace agent uses `Set` for some of the environment variable loading. I think this is worth discussing because I fear it might be a bad move to make Set private, even though in an ideal case it would be.

**Comment by gbbr** [comment]
Additionally, that `SetOverrides` method won't work with the legacy (A5) config loader as it is now. I strongly feel this is worth a discussion before considering removal. To my recall, during the last guild meeting that I was part of we were mostly in favor of leaving Set.

**Pull request merged**
